<a href="https://colab.research.google.com/github/olblinova/python_for_ds_tasks-is-available/blob/main/%22hw_10_4_apply%2C_groupby%2C_pivot_table_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
df = pd.read_csv('/content/drive/MyDrive/data/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [5]:
profit_by_branch = df.groupby('Branch')['gross income'].sum()

print(profit_by_branch)

print("Найприбутковіший філіал:", profit_by_branch.idxmax())

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64
Найприбутковіший філіал: C


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [6]:
most_profitable_branch = profit_by_branch.idxmax()

city = df.loc[df['Branch'] == most_profitable_branch, 'City'].iloc[0]

print(city)

Naypyitaw


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [7]:
invoices_by_category = pd.pivot_table(
    df,
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)

invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [8]:
branch = invoices_by_category['Electronic accessories'].idxmax()
count = invoices_by_category['Electronic accessories'].max()

print(f"Найбільше інвойсів з категорії 'Electronic accessories' у філії {branch}: {count} шт.")

Найбільше інвойсів з категорії 'Electronic accessories' у філії A: 60 шт.


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Який спосіь оплати використовують найчастіше клієнти в кожному філіалі?

In [10]:
payment_by_branch = pd.crosstab(df['Branch'], df['Payment'])

print(payment_by_branch)

Payment  Cash  Credit card  Ewallet
Branch                             
A         110          104      126
B         110          109      113
C         124           98      106


В філіалах А і В найчастіше сплачують з допомогою електронного гаманця, можливо вартує додати додатковий термінал для оплати для зручності. В філіалі С більше сплачують готівкою, можливо дадоти ще одну касу для прищвидшення розрахунків

Які категоріі товарів приносять найбільший середній чек з однієї покупки?

In [13]:
avg_profit_by_category = df.groupby('Product line')['gross income'].mean().sort_values(ascending=False)

print(avg_profit_by_category)

Product line
Home and lifestyle        16.030331
Sports and travel         15.812630
Health and beauty         15.411572
Food and beverages        15.365310
Electronic accessories    15.220597
Fashion accessories       14.528062
Name: gross income, dtype: float64


Найвищий середній прибуток приносить категорія Home and lifestyle — 16.03, можна провести рекламні заходи для збільшення попиту.Найнижчий — Fashion accessories — 14.53 - варто проаналізувати ціни, місцерозташування, можливо винести до каси, чи змінити викладку.

Які категорії товарів жінки та чоловіки купують частіше — чи справді жінки тягнуться до дорожчих категорій, чи просто купують частіше загалом?

In [14]:
df.groupby('Gender')['Product line'].value_counts(normalize=True).unstack().round(3) * 100

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Gender,,,,,,
Female,16.8,19.2,18.0,12.8,15.8,17.6
Male,17.2,16.4,16.8,17.6,16.2,15.6


In [15]:
df.groupby(['Product line', 'Gender'])['Total'].mean().round(2).unstack()

Gender,Female,Male
Product line,,
Electronic accessories,322.64,316.69
Fashion accessories,317.06,291.08
Food and beverages,368.57,273.50
Health and beauty,290.02,348.10
Home and lifestyle,380.21,294.14
Sports and travel,324.71,340.36


Загальна перевага жінок у середньому чеку тримається переважно на двох категоріях (їжа/напої та товари для дому), а не на всьому асортименті одразу. Преміум-пропозиції та бонусні програми для жінок можна порекоменндувати, оскільки чек тут уже високий і його можна нарощувати далі, розширити асортимент для чоловіків.